### PR TITLE
Issue/114/rewrite shear azimuthl averager

### PR DESCRIPTION
**Pull request by marina-ricci**
Pool request to include into masters the code to construct shear profiles from data. This contains : 

- clmm/polaveraging.py
- test/polavering.py

Some notebooks

### CODE REVIEW

**Comment by matthewkirby** [comment]
It looks like @mijinyoon is working in demo_polaraveraging and @marina-ricci is working in demo_of_polaraveraging.

There also appears to be a crazy file that is just a bunch of numbers .p in the examples directory. What even is that?

**Comment by matthewkirby** [comment]
This branch **should** pass travis. Then we can pull master into it, run tests locally and only then should we accept this.

**Comment by marina-ricci** [comment]
Ok, demo_polaraveraging is outdated, I will remove it. The file examples/9687686568.p is a pickle @matthewkirby created with DC2 data but since DC2 is not public I will remove it and update  demo_of_polaraveraging.ipynb to work with mock data instead of pickle. I do not understand the travis errors, I will have a look at that latter.

**Comment by matthewkirby** [comment]
Out of curiosity, why is like every other commit in here a merge?

**Comment by marina-ricci** [comment]
I do not understand your question.

**Comment by matthewkirby** [comment]
Screen Shot 2019-07-25 at 1 31 43 PM
Every other commit is a merge. When I see things like this I get a little concerned, thats all. If you look at the files changed in each one, the changes are the same for the commit before the merge and the merge. @aimalz @m-aguena do you guys have any thoughts on what is happening here?

File: tests/test_polaraveraging.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    #testing.assert_raises(UserWarning, pa._compute_radial_averages, radius=np.arange(1.,6.), g=np.arange(1.,6.), bins=[5,6,7]) 
+
+    
+
+def test_compute_theta_phi():
+    ra_l, dec_l = 161., 65.
+    ra_s, dec_s = np.array([-355., 355.]), np.array([-85., 85.])
+    rtol=1.e-7
+
+    # Test type on inputs - In retrospect, I don't think these tests really matter much as
```
Comment (lbaumo) [inline tests/test_polaraveraging.py]: Do we actually want to keep all of the commented out tests?  Does the math break in a predictable way?  It may be helpful to say why they are commented out

**Comment by aimalz** [comment]
@matthewkirby I am also perplexed by the merge commit issue and wonder if it's part of why I can't seem to remove those data files.  (Or does it appear to be working for you?)  I have tried [this](https://help.github.com/en/articles/removing-sensitive-data-from-a-repository) and [this](https://gist.github.com/magnetikonline/dd5837d597722c9c2d5dfa16d8efe5b9), which seem to work locally but seem to not solve the problem on GitHub, e.g. [here](https://github.com/LSSTDESC/CLMM/commit/eacb240e07d9ec9b8931515c8efd7cc3f1b81dc9#diff-bfebe34154a0dfd9fc7b447fc9ed74e9).  Regardless, it happened long enough ago that it doesn't matter if we fix it now or later, so it shouldn't stand in the way of approving this PR, but we should deal with it before making a release.

File: tests/test_polaraveraging.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    # ra1, ra2 = 0, 180.1
+    # testing.assert_allclose(pa._compute_theta_phi(), desired, rtol, err_msg="")
+    # ra1, ra2 = -180, 180
+    # testing.assert_allclose(pa._compute_theta_phi(), desired, rtol, err_msg="")
+    # dec1, dec2 = 90, -90
+    # testing.assert_allclose(pa._compute_theta_phi(), desired, rtol, err_msg="")
+
+
+def test_compute_shear():
+    # the numbers on this test must be revise, output values were extracted from the first run of the code
```
Comment (lbaumo) [inline tests/test_polaraveraging.py]: do we still need to revise these numbers?

Comment (marina-ricci) [reply tests/test_polaraveraging.py]: No. Lines 180 to 188 can be deleted.

File: clmm/polaraveraging.py [hunk 3]
```diff
@@ -2,5 +2,478 @@
...
+    g_x = - g_1 * \sin(2\phi) + g_2 * \cos(2\phi)
+    g_t = - (g_1 * \cos(2\phi) + g_2 * \sin(2\phi)) [cf. eqs. 7-8 of Schrabback et al. 2018, arXiv:1611.03866]
+    """
+    theta, phi = _compute_theta_phi(ra_l, dec_l, ra_s, dec_s, sky=sky)
+    g_t = _compute_g_t(g1, g2, phi)
+    g_x = _compute_g_x(g1, g2, phi)
+    return theta, g_t, g_x
+
+
+def _theta_units_conversion(theta, units, z_cl=None, cosmo=None, cosmo_object_type="astropy"):
```
[the remark this answers is not in the document]
Comment (m-aguena) [reply clmm/polaraveraging.py]: I just update this function for this simplicity, @marina-ricci can you test if it is working as it should?

File: clmm/polaraveraging.py [hunk 4]
```diff
@@ -2,5 +2,478 @@
...
+
+def _compute_g_x(g1, g2, phi):
+    r"""Computes cross shear for each source in galaxy catalog
+
+    Parameters
+    ----------
+    g1, g2,: array_like, float
+        ra and dec of the lens (l) and source (s)  in decimal degrees
+    phi: array_like, float
+        As defined in comput_theta_phi
```
[the remark this answers is not in the document]
Comment (marina-ricci) [reply clmm/polaraveraging.py]: Line 244 is completely wrong. I changed it.

File: clmm/polaraveraging.py [hunk 5]
```diff
@@ -2,5 +2,478 @@
...
+    r"""Computes the tangential shear for each source in the galaxy catalog
+
+    Add extended description
+
+    Parameters
+    ----------
+    g1, g2 : array_like, float
+        Ellipticity or shear for each source in the galaxy catalog
+    phi: array_like, float
+        As defined in comput_theta_phi (readdress this one)
```
[the remark this answers is not in the document]
Comment (marina-ricci) [reply clmm/polaraveraging.py]: I modify a bit line 214 and 246. However we still miss the full definition of phi. I let WL experts to fill that one.

File: clmm/polaraveraging.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (marina-ricci) [reply clmm/polaraveraging.py]: It works with ccl and astropy as expected.

File: clmm/polaraveraging.py [hunk 6]
```diff
@@ -2,5 +2,478 @@
...
+    g_x = _compute_g_x(g1, g2, phi)
+    return theta, g_t, g_x
+
+
+def _theta_units_conversion(theta, units, z_cl=None, cosmo=None, cosmo_object_type="astropy"):
+    """Converts theta from radian to whatever units specified in units
+
+    Parameters
+    ----------
+    theta : type???
```
[the remark this answers is not in the document]
Comment (marina-ricci) [reply clmm/polaraveraging.py]: Done

File: clmm/polaraveraging.py [hunk 7]
```diff
@@ -2,5 +2,485 @@
...
+        warnings.warn("Maximum radius is not within range of bins")
+    if np.amin(radius) < np.amin(bins):
+        warnings.warn("Minimum radius is not within the range of bins")
+
+    for i in range(len(bins)-1):
+        cond = (radius >= bins[i]) & (radius < bins[i+1])
+        index = np.where(cond)[0]
+        r_profile[i] = np.average(radius[index])
+        g_profile[i] = np.average(g[index])
+        if len(index) != 0:
```
[the remark this answers is not in the document]
Comment (matthewkirby) [reply clmm/polaraveraging.py]: Lets also move this check before r_profile and g_profile. `np.average([])` returns `nan` and throws a warning. We are already checking for this so lets just set it to `nan` manually just like gerr

File: tests/test_auxiliary_functions.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
+"""
+Tests for polaraveraging.py
```
Comment (matthewkirby) [inline tests/test_auxiliary_functions.py]: See, someone agreed that this should have been in a different place =P

File: tests/test_auxiliary_functions.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import clmm
+import clmm.polaraveraging as pa
+from numpy import testing
+import numpy as np
+
+
+def test_theta_units_conversion():
+    # tests for invaid input
+    testing.assert_raises(ValueError, pa._theta_units_conversion, np.pi, 'crazy units')
+    testing.assert_raises(ValueError, pa._theta_units_conversion, np.pi, 'Mpc')
```
Comment (matthewkirby) [inline tests/test_auxiliary_functions.py]: Lets move physical units and angular units conversions to diff functions. Split up tests as well

File: tests/test_auxiliary_functions.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    testing.assert_raises(ValueError, pa._theta_units_conversion, np.pi, 'Mpc', 0.5)
+
+    # tests for angular conversion
```
Comment (matthewkirby) [inline tests/test_auxiliary_functions.py]: We need to be sure to test every allowed conversion option. There were more than 4. For each, we should 
1. Test theta=0
2. Test theta=pi
3. Test theta=-pi
4. Test theta=48.48 (Just a random float)

File: tests/test_polaraveraging.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+Tests for polaraveraging.py
+"""
+import clmm
+import clmm.polaraveraging as pa
+from numpy import testing
+import numpy as np
+from astropy.table import Table
+import os
+
+def test_make_bins():
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: I am not reviewing these tests since I want to axe this function.

File: tests/test_polaraveraging.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+    testing.assert_raises(ValueError, pa.make_bins, rmin=1, rmax=10, n_bins=0, log_bins=False)
+    testing.assert_raises(TypeError, pa.make_bins, rmin=1, rmax=10, n_bins=9.9, log_bins=False)
+
+def test_compute_g_x():
+    data = np.array([[0.01, 0.02, 0.01], # g1
+                     [0.01, 0.02, 0.03], # g2
+                     [0.1, 1.2, 3.]]) # phi [rad]
+
+    # test that function works for scalar and vector input
+    assert(isinstance(pa._compute_g_x(*data[:,0]), float))
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: For this block of tests, call func once and save it to var. Then check that var however you want. A little cleaner than recalling the function every time.

File: tests/test_polaraveraging.py [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+    testing.assert_raises(ValueError, pa._compute_g_x, data[0], data[1,0], data[2,0])
+    testing.assert_raises(ValueError, pa._compute_g_x, data[0,0], data[1], data[2,0])
+    
+    # test for input range
+    testing.assert_raises(ValueError, pa._compute_g_x, 0.1, 0.1, -3.15)
+    testing.assert_raises(ValueError, pa._compute_g_x, 0.1, 0.1, 2.*np.pi+0.1)
+    testing.assert_almost_equal(pa._compute_g_x(0.1, 0.1, 0.), pa._compute_g_x(0.1, 0.1, np.pi))
+
+    
+    # test for reasonable values
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: These are corner cases not reasonable values.

File: tests/test_polaraveraging.py [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+    testing.assert_almost_equal(pa._compute_g_x(100., 0., 0.), 0.0)
+    testing.assert_almost_equal(pa._compute_g_x(0., 100., np.pi/4.), 0.0)
+    testing.assert_almost_equal(pa._compute_g_x(0., 0., 0.3), 0.)
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: Lets add some reasonable value tests. The math to compute the cross shear is super simple. We can choose same reasonable values, break out a calculator and compute the expectation. Lets do this for two tuples of phi, g1, g2

File: tests/test_polaraveraging.py [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+
+def test_compute_g_t():
+    data = np.array([[0.01, 0.02, 0.01], # g1
+                     [0.01, 0.02, 0.03], # g2
+                     [0.1, 1.2, 3.]]) # phi [rad]
+
+    # test that function works for scalar and vector input
+    # testing.assert(isinstance(float, pa._compute_g_t(*(data[:,0]))))
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: Why are these commented?

File: tests/test_polaraveraging.py [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+    testing.assert_raises(ValueError, pa._compute_g_x, 0.1, 0.1, 2.*np.pi+0.1)
+    testing.assert_almost_equal(pa._compute_g_x(0.1, 0.1, 0.), pa._compute_g_x(0.1, 0.1, np.pi))
+
+    
+    # test for reasonable values
+    testing.assert_almost_equal(pa._compute_g_x(100., 0., 0.), 0.0)
+    testing.assert_almost_equal(pa._compute_g_x(0., 100., np.pi/4.), 0.0)
+    testing.assert_almost_equal(pa._compute_g_x(0., 0., 0.3), 0.)
+
+def test_compute_g_t():
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: Same general comments as g_x

File: tests/test_polaraveraging.py [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+def test_compute_radial_averages():
+    #testing input types
+    testing.assert_raises(TypeError, pa._compute_radial_averages, radius="glue", g=10, bins=[np.arange(1.,16.)])
+    testing.assert_raises(TypeError, pa._compute_radial_averages, radius=np.arange(1.,10.), g="glue", bins=[np.arange(1.,16.)])  
+    testing.assert_raises(TypeError, pa._compute_radial_averages, radius=np.arange(1.,10.), g=np.arange(1.,10.), bins='glue') 
+
+    #want radius and g to have same number of entries
+    testing.assert_raises(TypeError, pa._compute_radial_averages, radius=np.arange(1.,10.), g=np.arange(1.,7.), bins=[np.arange(1.,16.)])
+
+    #want binning to encompass entire radial range
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: Why are these commented? Clarify comment to "ensure we throw a warning if gals fall out of bins range" or something.

File: tests/test_polaraveraging.py [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+    testing.assert_raises(ValueError, pa._compute_g_t, 0.1, 0.1, 2.*np.pi+0.1)
+    testing.assert_almost_equal(pa._compute_g_t(0.1, 0.1, 0.), pa._compute_g_t(0.1, 0.1, np.pi))
+    
+    # test for reasonable values
+    testing.assert_almost_equal(pa._compute_g_t(100., 0., 0.), -100.0)
+    testing.assert_almost_equal(pa._compute_g_t(0., 100., np.pi/4.), -100.0)
+    testing.assert_almost_equal(pa._compute_g_t(0., 0., 0.3), 0.)
+
+    
+def test_compute_radial_averages():
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: This function has no tests to ensure the function is doing what we want! We need to make sure we test this! This is the main point of unit tests!!! Odds are if we introduced a bug that caused any of these tests to fail, the code would crash anyways.

We need to test all of the allowed functionality, to 0th order, we at least need to test 
1. Pass in a list of numbers, bin edges -> compare output to the manually tabulated solution
2. Pass in a list of numbers -> compare output in default bins to manually tabulated solution
3. We should probably also test that things work when it performs unit conversions as well.

File: tests/test_polaraveraging.py (hunk 1 above)
Comment (matthewkirby) [reply tests/test_polaraveraging.py]: It looks like these are testing what happens if you pass in the wrong data type. In reality though, if you try to do `'0.5'+.5` your code is going to crash so these tests aren't really needed.

File: tests/test_polaraveraging.py [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+    testing.assert_raises(ValueError, pa._compute_theta_phi, -365., dec_l, ra_s, dec_s, 'flat')
+    testing.assert_raises(ValueError, pa._compute_theta_phi, 365., dec_l, ra_s, dec_s, 'flat')
+    testing.assert_raises(ValueError, pa._compute_theta_phi, ra_l, 95., ra_s, dec_s, 'flat')
+    testing.assert_raises(ValueError, pa._compute_theta_phi, ra_l, -95., ra_s, dec_s, 'flat')
+    testing.assert_raises(ValueError, pa._compute_theta_phi, ra_l, dec_l, ra_s-10., dec_s, 'flat')
+    testing.assert_raises(ValueError, pa._compute_theta_phi, ra_l, dec_l, ra_s+10., dec_s, 'flat')
+    testing.assert_raises(ValueError, pa._compute_theta_phi, ra_l, dec_l, ra_s, dec_s-10., 'flat')
+    testing.assert_raises(ValueError, pa._compute_theta_phi, ra_l, dec_l, ra_s, dec_s+10., 'flat')
+
+        # Test outputs for reasonable values with flat sky
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: extra tab

File: tests/test_polaraveraging.py [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+    ra_l, dec_l = 161.32, 51.49
+    ra_s, dec_s = np.array([161.29, 161.34]), np.array([51.45, 51.55])
+    thetas, phis = pa._compute_theta_phi(ra_l, dec_l, ra_s, dec_s, 'flat')
+    testing.assert_allclose(thetas, np.array([0.00077050407583119666, 0.00106951489719733675]), rtol=rtol,
+                            err_msg="Reasonable values with flat sky not matching to precision for theta")
+    testing.assert_allclose(phis, np.array([-1.13390499136495481736, 1.77544123918164542530]), rtol=rtol,
+                            err_msg="Reasonable values with flat sky not matching to precision for phi")
+
+    # Test that flat sky remains the default
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: these arent needed if we change func to be only flatsky

File: tests/test_polaraveraging.py [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+    thetas2, phis2 = pa._compute_theta_phi(ra_l, dec_l, ra_s, dec_s)
+    testing.assert_allclose(thetas2, thetas, rtol=rtol, err_msg="Theta for default sky value not in agreement for precision")
+    testing.assert_allclose(phis2, phis, rtol=rtol, err_msg="Phi for default sky value not in agreement for precision")
+
+    # Test outputs for reasonable values with curved sky - Not implemented in the code!
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: remove these. we arent supporting curved sky yet. They can be added (with a lot more tests) when we implement taht function

File: tests/test_polaraveraging.py [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+                            rtol, err_msg="Failure when lens and a source share an RA")
+    testing.assert_allclose(pa._compute_theta_phi(ra_l, dec_l, ra_s, np.array([51.49, 51.55])),
+                            [[0.00032601941539388962, 0.00106951489719733675], [0.00000000000000000000, 1.77544123918164542530]],
+                            rtol, err_msg="Failure when lens and a source share a DEC")
+    # l/s at same ra AND dec
+    testing.assert_raises(ValueError, pa._compute_theta_phi, ra_l, dec_l, np.array([161.32, 161.34]), np.array([51.49, 51.55]))
+
+    #testing.assert_warns(UserWarning,pa._compute_theta_phi, ra_l, dec_l, np.array([151.32, 161.34]), np.array([41.49, 51.55]),'flat')
+    
+    #testing.assert_allclose(pa._compute_theta_phi(ra_l, dec_l, np.array([161.32, 161.34]), np.array([51.49, 51.55])),
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: Why is this commented?

File: tests/test_polaraveraging.py [hunk 23]
```diff
@@ -0,0 +1,225 @@
...
+    testing.assert_allclose(pa._compute_theta_phi(ra_l, dec_l, np.array([161.32, 161.34]), dec_s),
+                            [[0.00069813170079771690, 0.00106951489719733675], [-1.57079632679489655800, 1.77544123918164542530]],
+                            rtol, err_msg="Failure when lens and a source share an RA")
+    testing.assert_allclose(pa._compute_theta_phi(ra_l, dec_l, ra_s, np.array([51.49, 51.55])),
+                            [[0.00032601941539388962, 0.00106951489719733675], [0.00000000000000000000, 1.77544123918164542530]],
+                            rtol, err_msg="Failure when lens and a source share a DEC")
+    # l/s at same ra AND dec
+    testing.assert_raises(ValueError, pa._compute_theta_phi, ra_l, dec_l, np.array([161.32, 161.34]), np.array([51.49, 51.55]))
+
+    #testing.assert_warns(UserWarning,pa._compute_theta_phi, ra_l, dec_l, np.array([151.32, 161.34]), np.array([41.49, 51.55]),'flat')
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: Why commented out?

File: tests/test_polaraveraging.py [hunk 24]
```diff
@@ -0,0 +1,225 @@
...
+    testing.assert_allclose(pa._compute_theta_phi(-180, dec_l, np.array([180.1,-90]), dec_s),
+                            [[2.36986569, 0.97805881], [-2.94587036e-04,  3.14052196e+00]],
+                            rtol, err_msg="Failure when ra_l and ra_s are the same but one is defined negative")     
+    
+    testing.assert_allclose(pa._compute_theta_phi(ra_l, 90, ra_s, np.array([51.45,-90])),
+                            [[0.67282443, 3.14159265], [-1.57079633, -1.57079633]],
+                            rtol, err_msg="Failure when dec_l and dec_s are separated by 180 deg")  
+
+
+def test_compute_shear():
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: This just looks kinda messy. Only need one simple test here. Pass in position of lens, position of 1 source, compare outputs to numbers you calculated on your calculator.

File: tests/test_polaraveraging.py [hunk 25]
```diff
@@ -0,0 +1,225 @@
...
+    ra_s, dec_s = np.array([161.29, 161.34]), np.array([51.45, 51.55])
+    testing.assert_allclose(
+        pa._compute_shear(ra_l, dec_l, ra_s, dec_s, g1, g2, sky='flat'),
+        np.array([[0.00077050407583119666, 0.00106951489719733675],
+            [ 1.408693e+00,  1.315362e+00],
+            [ 1.248319e-01, -5.194458e-01]]
+            ),
+        rtol=rtol)
+
+def test_gc_wrapper():
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: These are testing the monkey-patched versions of the functions. We can just axe this functions and move these tests to their appropriate functions. i.e. We should be testing that `compute_shear(Cluster, stuff)` works and `Clusters().compute_shear()` works together in the tests for `compute_shear()`.

In the above example, we can test `compute_shear(Cluster, stuff)` extensively and then just test `compute_shear(Cluster, stuff) == Cluster().compute_shear(stuff)`

File: tests/test_polaraveraging.py [hunk 26]
```diff
@@ -0,0 +1,225 @@
+"""
```
Comment (matthewkirby) [inline tests/test_polaraveraging.py]: No tests for

- [ ] '_astropy_to_CCL_cosmo_object', (I think this should be removed anyways)
- [ ]  'make_shear_profile',
- [ ]  'plot_profiles', (I think this should be moved to a plotting submodule)

**Comment by matthewkirby** [comment]
So I reviewed the code and in an effort to get this merged went ahead and made changes. The vast majority of the things I noted were style choices or small algorithm choices. I was hoping to make some fairly large changes but I am going to open a vote on those. See the slack channel for this discussion. We are really close to merging.

Of note, I changed a chunk of the backend api, this shouldn't affect the examples but I have not yet checked it.

**Comment by marina-ricci** [comment]
> So I reviewed the code and in an effort to get this merged went ahead and made changes. The vast majority of the things I noted were style choices or small algorithm choices. I was hoping to make some fairly large changes but I am going to open a vote on those. See the slack channel for this discussion. We are really close to merging.
> 
> Of note, I changed a chunk of the backend api, this shouldn't affect the examples but I have not yet checked it.

**Comment by matthewkirby** [comment]
@aimalz @cavestruz @marina-ricci I went ahead and finished this today. The only thing that I did not do is check the example notebooks (I made API changes so its possible that they are slightly broken, but it would only be adjusting function or keyword names). Marina and I extensively review and polaraveraging.py earlier so the new code that should be looked at is `utils.py` and `test_utils.py`. Let's get this merged on into master before hack day! =)

**Comment by matthewkirby** [comment]
100000000% make sure you click "Squash and Merge" when you merge this! Its 970 commits long!
